port/unix/omrsysinfo.c

```diff
@@ -4525,69 +4505,118 @@ omrsysinfo_cgroup_subsystem_iterator_hasNext(struct OMRPortLibrary *portLibrary,
 		subsystemMetricMap = omrCgroupCpusetMetricMap;
 		break;
 	default:
 		rc = OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_UNAVAILABLE;
+		state->count += 1;
 		goto _end;
 	}
 	subsystemMetricMapElement = &subsystemMetricMap[state->count];
 	currentElement = subsystemMetricMapElement->metricInfoElementList;
 	if (NULL == state->fileContent) {
-		if ((SINGLE_CGROUP_METRIC == subsystemMetricMapElement->metricElementsCount) && (NULL != currentElement->metricKeyInFile)) {
-			isSingleLineKeyValue = TRUE;
-		}
-		rc = readCgroupMetricFromFile(portLibrary, state->subsystemid, subsystemMetricMapElement->metricFileName, currentElement->metricKeyInFile, metricElement->value, &(*state).fileContent, isSingleLineKeyValue, sizeRef);
-		if (0 == rc) {
-			if(NULL == state->fileContent) {
-				state->multiLineCounter = SINGLE_CGROUP_METRIC;
-			} else {
-				state->multiLineCounter = 0;
-			}
-		} else {
+		rc = readCgroupMetricFromFile(portLibrary, state->subsystemid, subsystemMetricMapElement->metricFileName, currentElement->metricKeyInFile, &(state->fileContent), metricElement->value);
+
+		/* In error condition or single metric condition we increment the counter to continue to next metric */
+		if ((0 != rc) || (NULL == state->fileContent)) {
+			state->fileContent = NULL;
+			state->count += 1;
 			goto _end;
 		}
-		
-	}
-	if (state->multiLineCounter < subsystemMetricMapElement->metricElementsCount) {
-		char *tempBuff = portLibrary->mem_allocate_memory(portLibrary, 1024, OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
-		char *part = "";
-		char *saveptr = NULL;
-		currentElement += state->multiLineCounter;
-		strcpy(tempBuff, state->fileContent);
-		part = strtok_r(tempBuff, "\n", &saveptr);
-	  	do {
-			if(0 == strncmp(part, currentElement->metricKeyInFile, strlen(currentElement->metricKeyInFile))) {
-				part += strlen(currentElement->metricKeyInFile);
-				strcpy(metricElement->value, ++part);
-				rc = 0;
-				portLibrary->mem_free_memory(portLibrary, tempBuff);
-				state->multiLineCounter = state->multiLineCounter + 1;
-				goto _end;
+
+		state->fileMetricCounter = 0;
+	}
+
+	if (state->fileMetricCounter < subsystemMetricMapElement->metricElementsCount) {
+		currentElement += state->fileMetricCounter;
+		int32_t metricKeyLen = strlen(currentElement->metricKeyInFile);
+		BOOLEAN isMetricFound = FALSE;
+		int32_t lineStart = 0;
+		int32_t lineSpace = 0;
+		int32_t lineEnd = 0;
+		int32_t index = 0;
+		char *tempPtr = state->fileContent;
+		while (index < strlen(state->fileContent)) {
+			if (tempPtr[index] == ' ') {
+				lineSpace = index;
+			} else if (tempPtr[index] == '\n') {
+				lineEnd = index;
+				if (0 == strncmp(&tempPtr[lineStart], currentElement->metricKeyInFile, metricKeyLen)) {
+					memset(metricElement->value, 0, MAX_LINE_LENGTH);
+					strncpy(metricElement->value, &tempPtr[lineSpace + 1], lineEnd - lineSpace);
+					isMetricFound = TRUE;
+					rc = 0;
+					break;
+				}
+				lineStart = index + 1;
```

Comment:
I think this would be much clearer if written using pointers to search the string and would avoid some of the signed / unsigned comparisons.  Something along the lines of:

```c
char *start = state->fileContent;
char *current = state->fileContent;
char *end = state->fileContent + strlen(state->fileContent) - metricKeyLen;  /* Must ensure there is at least metric worth of data */
while (current < end) {
    if (0 == strncmp(start, currentElement->metricKeyInFile, metricKeyLen)) {
        char *newLine = current;
        current += metricKeyLen;
        /* advance past any whitespace */
        while ((current != end) &&  (' ' == *current)) {
            current += 1;
        }
        /* find the newline */
        newLine = current;       
        while ((newLine != end) && ('\n' != *newLine)) {
            newLine += 1;
        }
        strncpy(metricElement->value, current, newLine - current);
    }
    current += 1;
}
```

Reply:
@DanHeidinga Thank you for suggesting this method, I have implemented it with minor changes. Please let me know any changes required.